uaclient/clouds/gcp.py

```diff
@@ -49,19 +62,78 @@ def is_viable(self) -> bool:
-            .get("license_id", [])
+        Checks list of licenses from the GCP metadata.
+        """
+        series = util.get_platform_info()["series"]
+        if series not in GCP_LICENSES:
+            return False
+
+        licenses, headers = util.readurl(
+            LICENSES_URL, headers={"Metadata-Flavor": "Google"}
         )
+        license_ids = [license["id"] for license in licenses]
+        self.etag = headers.get("ETag", None)
+
+        return GCP_LICENSES[series] in license_ids
+
+    def should_poll_for_license(self) -> bool:
+        series = util.get_platform_info()["series"]
+        if series not in GCP_LICENSES:
+            LOG.info("This series isn't supported for GCP auto-attach.")
+            return False
+
+        return True
+
+    def get_polling_fn(self) -> Optional[Callable]:
+        def gcp_polling_fn():
+            LOG.debug("gcp_polling_fn started")
+            done = False
+            while not done:
+                license_ids = []
+                try:
+                    # TODO: should we have a timeout? configurable? default?
+                    # https://cloud.google.com/compute/docs/metadata/querying-metadata#settingtimeouts  # noqa
+                    url = LICENSES_URL + WAIT_FOR_CHANGE
+                    if self.etag:
+                        url += LAST_ETAG.format(self.etag)
+                    licenses, headers = util.readurl(
+                        url, headers={"Metadata-Flavor": "Google"}
+                    )
+                    license_ids = [
+                        license.get("id", "") for license in licenses
+                    ]
+                    self.etag = headers.get("ETag", None)
+                except HTTPError as e:
+                    LOG.error(e)
+                    if e.code == 400:
+                        LOG.debug("Got 400 from metadata. Cancelling polling")
+                        done = True
+                    else:
+                        LOG.debug("waiting 10 minutes before trying again")
+                        time.sleep(600)
+                except Exception as e:
+                    LOG.exception(e)
+                    LOG.debug("waiting 10 minutes before trying again")
+                    time.sleep(600)
+
+                if license_ids:
+                    series = util.get_platform_info()["series"]
+                    if GCP_LICENSES[series] in license_ids:
+                        LOG.info("pro license found. auto-attaching")
+                        try:
+                            with lock.SpinLock(
+                                cfg=self.cfg,
+                                lock_holder="ua.clouds.gcp.gcp_polling_fn",
+                            ):
+                                actions.auto_attach(self.cfg, self)
+                        except Exception as e:
+                            lock.clear_lock_file_if_present()
+                            # TODO: should we retry here?
```

Comment:
I guess we could retry a couple times

Reply:
+1

Reply:
@renanrodrigo @lucasmoura 
Just to double check: because of the SpinLock, this will already try to acquire the lock several times.
Do you think we should retry more on top of that? (in the event of a exception != LockHeldError)